kollect-core/src/main/kotlin/kollect/Kollect.kt

```diff
@@ -1,175 +1,399 @@
 import arrow.data.NonEmptyList
-import arrow.data.fix
-import arrow.data.k
-import arrow.free.Free
-import arrow.free.FreePartialOf
-import arrow.free.fix
+import arrow.effects.Promise
+import arrow.effects.Ref
+import arrow.effects.typeclasses.Concurrent
 import arrow.higherkind
-import arrow.instance
 import arrow.typeclasses.Applicative
+import arrow.typeclasses.Monad
+import kollect.typeclasses.Timer
+import java.util.concurrent.TimeUnit
 
-typealias Kollect<A> = Free<ForKollectOp, A>
-
-abstract class NoStackTrace : Throwable() {
-    override fun fillInStackTrace(): Throwable = this
-}
-
-sealed class KollectError : NoStackTrace() {
-    abstract val env: Env
-}
-
-data class NotFound(override val env: Env, val request: KollectOp.KollectOne<Any, Any>) : KollectError()
-data class MissingIdentities(override val env: Env, val missing: Map<DataSourceName, List<Any>>) : KollectError()
-
-data class UnhandledException(override val env: Env, val err: Throwable) : KollectError()
-
-interface KollectRequest
+@higherkind
+sealed class Kollect<F, A> : KollectOf<F, A> {
 
-interface KollectQuery<I : Any, A> : KollectRequest {
-    abstract fun dataSource(): DataSource<I, A>
-    abstract fun identities(): NonEmptyList<I>
-}
+    abstract val run: arrow.Kind<F, KollectResult<F, A>>
 
-/**
- * Primitive operations in the Kollect Free monad.
- */
-@higherkind
-sealed class KollectOp<out A> : KollectOpOf<A> {
-    data class Thrown<A>(val err: Throwable) : KollectOp<A>()
-    data class Join<A, B>(val fl: Kollect<A>, val fr: Kollect<B>) : KollectOp<Tuple2<A, B>>()
-    data class Concurrent(val queries: NonEmptyList<KollectQuery<Any, Any>>) : KollectOp<InMemoryCache>(), KollectRequest
-    data class KollectOne<I : Any, A>(val id: I, val ds: DataSource<I, A>) : KollectOp<A>(), KollectQuery<I, A> {
-        override fun dataSource(): DataSource<I, A> = ds
-        override fun identities(): NonEmptyList<I> = NonEmptyList.just(id)
-    }
+    data class Unkollect<F, A>(override val run: arrow.Kind<F, KollectResult<F, A>>) : Kollect<F, A>()
 
-    data class KollectMany<I : Any, A>(val ids: NonEmptyList<I>, val ds: DataSource<I, A>) : KollectOp<List<A>>(), KollectQuery<I, A> {
-        override fun dataSource(): DataSource<I, A> = ds
-        override fun identities(): NonEmptyList<I> = ids
+    fun <B> map(MF: Monad<F>, f: (A) -> B): Kollect<F, B> =
+            Kollect.Unkollect(MF.binding {
+                val kollect = run.bind()
+                val result = when (kollect) {
+                    is KollectResult.Done -> KollectResult.Done<F, B>(f(kollect.x))
+                    is KollectResult.Blocked -> KollectResult.Blocked(kollect.rs, kollect.cont.map(MF, f))
+                    is KollectResult.Throw -> KollectResult.Throw(kollect.e)
+                }
+                result
+            })
+
+    fun <I, B> product(MF: Monad<F>, fb: Kind<KollectPartialOf<F>, B>): Kollect<F, Tuple2<A, B>> =
+            Unkollect(MF.binding {
```

Comment:
Consider using `flatMap` instead since it only binds once. The resulting code will have same indentation without the current cost we pay for `binding`.

Reply:
Yep makes sense. Done now 👍